pkg/controllers/node/termination/terminator/eviction.go

```diff
@@ -71,117 +75,107 @@ func IsNodeDrainError(err error) bool {
-		Named("eviction-queue").
-		WatchesRawSource(singleton.Source()).
-		Complete(singleton.AsReconciler(q))
+		Named(q.Name()).
+		WatchesRawSource(source.Channel(q.source, handler.TypedFuncs[*corev1.Pod, reconcile.Request]{
+			GenericFunc: func(_ context.Context, e event.TypedGenericEvent[*corev1.Pod], queue workqueue.TypedRateLimitingInterface[reconcile.Request]) {
+				queue.Add(reconcile.Request{
+					NamespacedName: client.ObjectKeyFromObject(e.Object),
+				})
+			},
+		})).
+		WithOptions(controller.Options{
+			RateLimiter: workqueue.NewTypedMaxOfRateLimiter[reconcile.Request](
+				workqueue.NewTypedItemExponentialFailureRateLimiter[reconcile.Request](evictionQueueBaseDelay, evictionQueueMaxDelay),
+				&workqueue.TypedBucketRateLimiter[reconcile.Request]{Limiter: rate.NewLimiter(rate.Limit(100), 1000)},
+			),
+			MaxConcurrentReconciles: 100,
+		}).
+		Complete(reconcile.AsReconciler(m.GetClient(), q))
 }
 
 // Add adds pods to the Queue
-func (q *Queue) Add(node *corev1.Node, pods ...*corev1.Pod) {
-	q.mu.Lock()
-	defer q.mu.Unlock()
+func (q *Queue) Add(pods ...*corev1.Pod) {
+	q.Lock()
+	defer q.Unlock()
 
 	for _, pod := range pods {
-		qk := NewQueueKey(pod, node.Spec.ProviderID)
+		qk := NewQueueKey(pod)
 		if !q.set.Has(qk) {
 			q.set.Insert(qk)
-			q.TypedRateLimitingInterface.Add(qk)
+			q.source <- event.TypedGenericEvent[*corev1.Pod]{Object: pod}
 		}
 	}
 }
 
-func (q *Queue) Has(node *corev1.Node, pod *corev1.Pod) bool {
-	q.mu.Lock()
-	defer q.mu.Unlock()
+func (q *Queue) Has(pod *corev1.Pod) bool {
+	q.Lock()
+	defer q.Unlock()
 
-	return q.set.Has(NewQueueKey(pod, node.Spec.ProviderID))
+	return q.set.Has(NewQueueKey(pod))
 }
 
-func (q *Queue) Reconcile(ctx context.Context) (reconcile.Result, error) {
-	ctx = injection.WithControllerName(ctx, "eviction-queue")
-	// Check if the queue is empty. client-go recommends not using this function to gate the subsequent
-	// get call, but since we're popping items off the queue synchronously, there should be no synchonization
-	// issues.
-	if q.TypedRateLimitingInterface.Len() == 0 {
-		return reconcile.Result{RequeueAfter: 1 * time.Second}, nil
-	}
-	// Get pod from queue. This waits until queue is non-empty.
-	item, shutdown := q.TypedRateLimitingInterface.Get()
-	if shutdown {
-		return reconcile.Result{}, fmt.Errorf("EvictionQueue is broken and has shutdown")
-	}
-
-	defer q.TypedRateLimitingInterface.Done(item)
+func (q *Queue) Reconcile(ctx context.Context, pod *corev1.Pod) (reconcile.Result, error) {
+	ctx = injection.WithControllerName(ctx, q.Name())
 
+	if !q.Has(pod) {
+		//This is a different pod than in the queue, we should exit without evicting
+		return reconcile.Result{}, nil
+	}
 	// Evict the pod
-	if q.Evict(ctx, item) {
-		q.TypedRateLimitingInterface.Forget(item)
-		q.mu.Lock()
-		q.set.Delete(item)
-		q.mu.Unlock()
-		return reconcile.Result{RequeueAfter: singleton.RequeueImmediately}, nil
+	if err := q.Evict(ctx, pod); err != nil {
```

Comment:
If we handle errors like this, we are going to be printing a lot of errors as we encounter different pods that have PDBs that are blocking their eviction -- I think we need to avoid returning an error here and handle the Requeuing in a different way with a rate limiter OR with RequeueAfter